// raml-actions/dist/index.js
"use strict";
var contextActions = require("./actionManagement/contextActions");
var contextActionsImpl = require("./actionManagement/contextActionsImpl");
var sharedStateCalculator = require("./actionManagement/sharedASTStateCalculator");
var contextMenuImpl = require("./actionManagement/contextMenuImpl");
var actionInitializer = require("./actions/actionInitializer");
exports.uiActions = require("./actions/uiActions");
exports.standardActions = require("./actions/standardActions");
exports.parser = sharedStateCalculator.parser;
exports.TARGET_RAML_EDITOR_NODE = contextActions.TARGET_RAML_EDITOR_NODE;
exports.TARGET_RAML_TREE_VIEWER_NODE = contextActions.TARGET_RAML_TREE_VIEWER_NODE;
function addAction(action) {
    contextActionsImpl.addAction(action);
}
exports.addAction = addAction;
function addSimpleAction(name, category, target, onClick, shouldDisplay) {
    contextActionsImpl.addSimpleAction(name, category, target, onClick, shouldDisplay);
}
exports.addSimpleAction = addSimpleAction;
function calculateCurrentActions(target) {
    return contextActionsImpl.calculateCurrentActions(target);
}
exports.calculateCurrentActions = calculateCurrentActions;
function getCategorizedActionLabel(action) {
    return contextActionsImpl.getCategorizedActionLabel(action);
}
exports.getCategorizedActionLabel = getCategorizedActionLabel;
var CommonASTStateCalculator = (function () {
    function CommonASTStateCalculator() {
        this.contextCalculationStarted = function () {
            sharedStateCalculator.generalASTStateCalculator.contextCalculationStarted();
        };
        this.contextCalculationFinished = function () {
            sharedStateCalculator.generalASTStateCalculator.contextCalculationFinished();
        };
    }
    CommonASTStateCalculator.prototype.calculate = function () {
        return null;
    };
    CommonASTStateCalculator.prototype.getGeneralState = function () {
        return sharedStateCalculator.generalASTStateCalculator.calculate();
    };
    CommonASTStateCalculator.prototype.getEditor = function () {
        return sharedStateCalculator.generalASTStateCalculator.getEditor();
    };
    return CommonASTStateCalculator;
}());
exports.CommonASTStateCalculator = CommonASTStateCalculator;
function setEditorProvider(editorProvider) {
    sharedStateCalculator.setEditorProvider(editorProvider);
}
exports.setEditorProvider = setEditorProvider;
function setASTProvider(astProvider) {
    sharedStateCalculator.setASTProvider(astProvider);
}
exports.setASTProvider = setASTProvider;
function setASTModifier(astModifier) {
    sharedStateCalculator.setASTModifier(astModifier);
}
exports.setASTModifier = setASTModifier;
function registerContributor(contributor) {
    contextMenuImpl.registerContributor(contributor);
}
exports.registerContributor = registerContributor;
function calculateMenuItemsTree() {
    return contextMenuImpl.calculateMenuItemsTree();
}
exports.calculateMenuItemsTree = calculateMenuItemsTree;
function intializeStandardActions() {
    actionInitializer.intializeStandardActions();
}
exports.intializeStandardActions = intializeStandardActions;
function initializeActionBasedMenu(selector) {
    contextMenuImpl.initializeActionBasedMenu(selector);
}
exports.initializeActionBasedMenu = initializeActionBasedMenu;
