// raml-typesystem/dist/src/typesystem-interfaces.js
"use strict";
exports.REPEAT = "repeat";
exports.PARSE_ERROR = "parseError";
exports.TOP_LEVEL_EXTRA = "topLevel";
exports.DEFINED_IN_TYPES_EXTRA = "definedInTypes";
exports.USER_DEFINED_EXTRA = "USER_DEFINED";
exports.SOURCE_EXTRA = "SOURCE";
exports.SCHEMA_AND_TYPE_EXTRA = "SCHEMA";
exports.GLOBAL_EXTRA = "GLOBAL";
exports.HAS_FACETS = "HAS_FACETS";
exports.HAS_ITEMS = "HAS_ITEMS";
(function (MetaInformationKind) {
    MetaInformationKind[MetaInformationKind["Description"] = 0] = "Description";
    MetaInformationKind[MetaInformationKind["NotScalar"] = 1] = "NotScalar";
    MetaInformationKind[MetaInformationKind["DisplayName"] = 2] = "DisplayName";
    MetaInformationKind[MetaInformationKind["Usage"] = 3] = "Usage";
    MetaInformationKind[MetaInformationKind["Annotation"] = 4] = "Annotation";
    MetaInformationKind[MetaInformationKind["FacetDeclaration"] = 5] = "FacetDeclaration";
    MetaInformationKind[MetaInformationKind["CustomFacet"] = 6] = "CustomFacet";
    MetaInformationKind[MetaInformationKind["Example"] = 7] = "Example";
    MetaInformationKind[MetaInformationKind["Required"] = 8] = "Required";
    MetaInformationKind[MetaInformationKind["AllowedTargets"] = 9] = "AllowedTargets";
    MetaInformationKind[MetaInformationKind["Examples"] = 10] = "Examples";
    MetaInformationKind[MetaInformationKind["XMLInfo"] = 11] = "XMLInfo";
    MetaInformationKind[MetaInformationKind["Default"] = 12] = "Default";
    MetaInformationKind[MetaInformationKind["Constraint"] = 13] = "Constraint";
    MetaInformationKind[MetaInformationKind["Modifier"] = 14] = "Modifier";
    MetaInformationKind[MetaInformationKind["Discriminator"] = 15] = "Discriminator";
    MetaInformationKind[MetaInformationKind["DiscriminatorValue"] = 16] = "DiscriminatorValue";
})(exports.MetaInformationKind || (exports.MetaInformationKind = {}));
var MetaInformationKind = exports.MetaInformationKind;
