// atom-ui-lib/dist/fileDialogUtils.js
"use strict";
var remote = require("remote");
var dialog = remote.require("dialog");
function openFileDialogModal(title, defaultPath, filters) {
    var options = constructOptions(title, defaultPath, filters, ['openFile']);
    return dialog.showOpenDialog(options);
}
exports.openFileDialogModal = openFileDialogModal;
function openFileDialog(title, callBack, defaultPath, filters) {
    var options = constructOptions(title, defaultPath, filters, ['openFile']);
    dialog.showOpenDialog(options, function (resultPath) {
        if (resultPath)
            callBack(resultPath[0]);
    });
}
exports.openFileDialog = openFileDialog;
function openFolderDialogModal(title, createDirectory, defaultPath, filters) {
    var properties = ['openDirectory'];
    if (createDirectory)
        properties.push('createDirectory');
    var options = constructOptions(title, defaultPath, filters, properties);
    return dialog.showOpenDialog(options);
}
exports.openFolderDialogModal = openFolderDialogModal;
function openFolderDialog(title, callBack, createDirectory, defaultPath, filters) {
    var properties = ['openDirectory'];
    if (createDirectory)
        properties.push('createDirectory');
    var options = constructOptions(title, defaultPath, filters, properties);
    dialog.showOpenDialog(options, function (resultPath) {
        if (resultPath)
            callBack(resultPath[0]);
    });
}
exports.openFolderDialog = openFolderDialog;
function saveFileDialogModal(title, defaultPath, filters) {
    var options = constructOptions(title, defaultPath, filters, ['saveFile']);
    return dialog.showSaveDialog(options);
}
exports.saveFileDialogModal = saveFileDialogModal;
function constructOptions(title, defaultPath, filters, properties) {
    var options = {
        title: title,
        properties: properties,
    };
    if (defaultPath) {
        options.defaultPath = defaultPath;
    }
    if (filters) {
        options.filters = filters;
    }
    return options;
}
function getHome() {
    var home = process.env["HOME"];
    if (home) {
        return home;
    }
    var userProfile = process.env["USERPROFILE"];
    if (userProfile) {
        return userProfile;
    }
    var publicFolder = process.env["PUBLIC"];
    if (publicFolder) {
        return publicFolder;
    }
    var atomHome = process.env["ATOM_HOME"];
    if (atomHome) {
        return atomHome;
    }
    return "";
}
exports.getHome = getHome;
